[benchmark][cudagraph] Explicitly call aten.div with CUDA denominator for cudagraphs

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124804
* #124700
* __->__ #119729

aten.div's output device will be its numerator's device. so it is acceptable to do cuda / cpu type divisions. post grad passes operate only on graphs and can't handle runtime graph inputs. so we change user code to move inputs to cuda for cudagraph. this affects any graph that has cpu tensors as graph inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

xmfan committed:
Fix cpu division in benchmarks affecting compiled autograd + cudagraphs

[ghstack-poisoned]

eellison reviewed (commented):
This is good, but it would also be nice to expand 
https://github.com/pytorch/pytorch/blob/bd9db6a9c79c68eb032df78d9efcdd27088209ec/torch/_inductor/fx_passes/post_grad.py#L102 

to encompass this case. do you have a test that would demonstrate the issue ?

xmfan commented:
I have a repro/test to add: https://gist.github.com/xmfan/6f1bc7e487bf1bdb97ae467d23ec9f21. I added some logging statements to that function:

``` python
class ConstructorMoverPass:
    def __call__(self, graph: fx.Graph) -> None:
        target_devices = set()
        constructors = []

        for node in graph.nodes:
            breakpoint()
            """
            (Pdb) p self.get_node_device(node)
            device(type='cuda', index=0)
            (Pdb) p node.target
            <OpOverload(op='aten.div', overload='Tensor')>
            (Pdb) p node.kwargs
            {}
            """

            device = self.get_node_device(node)
            if device and device.type == self.target: # True
                target_devices.add(device)

            if not (
                isinstance(node.target, torch._ops.OpOverload)
                and node.target.namespace in ("prims", "aten")
            ): # False
                continue

            if not torch._subclasses.fake_tensor._is_tensor_constructor(node.target): # True
                continue 

            if not node.kwargs.get("device") == torch.device("cpu"): # True
                continue

            constructors.append(node)
```

At first, I thought that we needed the div node to be considered as a constructor on CPU. But I notice that it is also not appended in compiled forward, yet it has no cudagraph warning message (https://gist.github.com/xmfan/c8fe3b2c3e2dddae25883b74c402d44d)

xmfan committed:
Update on "Fix cpu division in benchmarks affecting compiled autograd + cudagraphs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_dynamo/testing.py:
```diff
@@ -103,7 +103,7 @@ def reduce_to_scalar_loss(out):
     if isinstance(out, torch.Tensor):
         # Mean does not work on integer tensors
-        return out.sum() / out.numel()
+        return out.sum() / torch.tensor(out.numel(), device=out.device)
```
Why is this not captured the graph itself instead of lifted as an input ?

```
    def forward(self, arg0_1: "f16[]cuda:0", arg1_1: "i64[]cpu", 
```

xmfan replied:
compiled autograd lifts all tensors as it traces the autograd nodes

xmfan committed:
Update on "[benchmark][cudagraph] Explicitly call aten.div with CUDA denominator for cudagraphs"


aten.div's output device will be its numerator's device. so it is acceptable to do cuda / cpu type divisions. post grad passes operate only on graphs and can't handle runtime graph inputs. so we change user code to move inputs to cuda for cudagraph. this affects any graph that has cpu tensors as graph inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jeanschmidt commented:
@pytorchbot revert -m "one PR in this stack seems to have broken linux pull cuda12 tests" -c ignoredsignal
